## **02_Data_EDA**

In dieser Dateien ... .

In [685]:
# Imports
import pandas as pd
import os
import numpy as np


# Profilingreport
from ydata_profiling import ProfileReport
import webbrowser
import warnings

warnings.filterwarnings("ignore", category=UserWarning, module="geopy")


In [686]:
# Alle Spalten eines Datensatzes anzeigen
pd.set_option('display.max_columns', None) 

### Daten einlesen

In [687]:
df = pd.read_csv('../data/input_data/property_sales_2004-2024.csv')

/var/folders/jf/k1y_jzxs5177jm_fgv__dnww0000gn/T/ipykernel_32773/3420834570.py:1: DtypeWarning: Columns (16,18,23) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/input_data/property_sales_2004-2024.csv')


In [688]:
# Ausgabe der ersten 5 Zeilen
df.head()

PropType        Taxkey           Address CondoProject  District    Nbhd  \
0  Residential  1.230322e+09    7144 N 38TH ST          NaN       1.0   600.0   
1  Residential  1.909905e+09    5346 N 52ND ST          NaN       1.0  1140.0   
2  Residential  2.441200e+09    4155 N 13TH ST          NaN       1.0  1620.0   
3  Residential  1.230292e+09  3832 W HELENA ST          NaN       1.0   600.0   
4  Residential  2.080590e+09    4862 N 39TH ST          NaN       1.0  1440.0   

      Style           Extwall  Stories  Year_Built  Nr_of_rms  Fin_sqft  \
0  Cape Cod  Aluminum / Vinyl      1.5      1959.0        0.0    1734.0   
1  Cape Cod             Brick      1.0      1951.0        0.0    1718.0   
2  Cape Cod             Brick      1.5      1938.0        0.0    1614.0   
3  Cape Cod  Aluminum / Vinyl      1.5      1960.0        0.0    1582.0   
4  Cape Cod  Aluminum / Vinyl      1.0      1954.0        0.0    1556.0   

   Units  Bdrms  Fbath  Hbath Lotsize Sale_date Sale_price  PropertyID  \
0    1.0    4.0    2.0    0.0    7000   2018-07     129000         NaN   
1    1.0    3.0    2.0    1.0    8100   2018-08      67000         NaN   
2    1.0    3.0    1.0    1.0    5040   2018-02     103000         NaN   
3    1.0    4.0    2.0    0.0    6000   2018-03     150000         NaN   
4    1.0    4.0    2.0    1.0    4920   2018-03      89900         NaN   

   taxkey  nbhd  Rooms FinishedSqft  
0     NaN   NaN    NaN          NaN  
1     NaN   NaN    NaN          NaN  
2     NaN   NaN    NaN          NaN  
3     NaN   NaN    NaN          NaN  
4     NaN   NaN    NaN          NaN

### Daten beschreiben

In [689]:
# Ausgabe der Datensatz-Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67635 entries, 0 to 67634
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PropType      67622 non-null  object 
 1   Taxkey        29899 non-null  float64
 2   Address       67629 non-null  object 
 3   CondoProject  10696 non-null  object 
 4   District      67631 non-null  float64
 5   Nbhd          29899 non-null  float64
 6   Style         67532 non-null  object 
 7   Extwall       53992 non-null  object 
 8   Stories       67446 non-null  float64
 9   Year_Built    67567 non-null  float64
 10  Nr_of_rms     29899 non-null  float64
 11  Fin_sqft      29899 non-null  float64
 12  Units         67633 non-null  float64
 13  Bdrms         65688 non-null  float64
 14  Fbath         66978 non-null  float64
 15  Hbath         62793 non-null  float64
 16  Lotsize       67632 non-null  object 
 17  Sale_date     67633 non-null  object 
 18  Sale_price    67623 non-nu

In [690]:
# Ausgabe der Anzahl der Zeilen und Spalten
print(f'Anzahl Zeilen: {df.shape[0]}')
print(f'Anzahl Spalten: {df.shape[1]}')

Anzahl Zeilen: 67635
Anzahl Spalten: 24


In [691]:
# Ausgabe der numerischen und kategorischen Variablen 
numericalVar = [col for col in df if df[col].dtype != 'object']
print(f"Numerische Variablen: {numericalVar}")

categoricalVar = [col for col in df if df[col].dtype == 'object']
print(f"Kategorische Variablen: {categoricalVar}")

Numerische Variablen: ['Taxkey', 'District', 'Nbhd', 'Stories', 'Year_Built', 'Nr_of_rms', 'Fin_sqft', 'Units', 'Bdrms', 'Fbath', 'Hbath', 'PropertyID', 'taxkey', 'nbhd', 'Rooms']
Kategorische Variablen: ['PropType', 'Address', 'CondoProject', 'Style', 'Extwall', 'Lotsize', 'Sale_date', 'Sale_price', 'FinishedSqft']


In [692]:
# Anzahl der NaN-Werte pro Spalte

# Anzhal NaN 
count_nan = df.isna().sum()
# Prozentsatz NaN
nan_percent = (count_nan / len(df)) * 100
# DataFrame
df_nan = pd.DataFrame({
    'Anzahl NaN': count_nan,
    'Prozent NaN_': nan_percent
}).sort_values(by='Anzahl NaN', ascending=False)
df_nan

Anzahl NaN  Prozent NaN_
CondoProject       56939     84.185703
Nr_of_rms          37736     55.793598
Nbhd               37736     55.793598
Taxkey             37736     55.793598
Fin_sqft           37736     55.793598
Rooms              31847     47.086568
FinishedSqft       30013     44.374954
nbhd               29958     44.293635
taxkey             29901     44.209359
PropertyID         29901     44.209359
Extwall            13643     20.171509
Hbath               4842      7.159015
Bdrms               1947      2.878687
Fbath                657      0.971391
Stories              189      0.279441
Style                103      0.152288
Year_Built            68      0.100540
PropType              13      0.019221
Sale_price            12      0.017742
Address                6      0.008871
District               4      0.005914
Lotsize                3      0.004436
Sale_date              2      0.002957
Units                  2      0.002957

In [693]:
# Ausgabe der statistischen Kennzahlen der numerischen Variablen
df.describe().T

count          mean           std         min           25%  \
Taxkey      29899.0  3.734459e+09  1.479836e+09  10021000.0  2.706017e+09   
District    67631.0  7.957623e+00  4.218837e+00         1.0  4.000000e+00   
Nbhd        29899.0  3.822750e+03  1.799399e+03        40.0  2.100000e+03   
Stories     67446.0  1.356923e+00  6.191259e-01         0.0  1.000000e+00   
Year_Built  67567.0  1.926929e+03  1.825797e+02         0.0  1.923000e+03   
Nr_of_rms   29899.0  6.835680e-01  1.642115e+00         0.0  0.000000e+00   
Fin_sqft    29899.0  2.535691e+03  9.259781e+03         0.0  1.063000e+03   
Units       67633.0  1.547957e+00  8.231289e+00         0.0  1.000000e+00   
Bdrms       65688.0  3.139615e+00  8.068394e+00         0.0  2.000000e+00   
Fbath       66978.0  1.375019e+00  7.127618e-01         0.0  1.000000e+00   
Hbath       62793.0  3.171054e-01  5.138513e-01         0.0  0.000000e+00   
PropertyID  37734.0  4.123111e+05  3.272643e+05     98422.0  1.588952e+05   
taxkey      37734.0  3.488834e+09  1.492266e+09  10011000.0  2.452054e+09   
nbhd        37677.0  3.373693e+03  1.819229e+03        40.0  1.840000e+03   
Rooms       35788.0  6.730077e+00  3.242651e+00         0.0  5.000000e+00   

                     50%           75%           max  
Taxkey      3.591432e+09  5.070828e+09  7.169999e+09  
District    8.000000e+00  1.100000e+01  1.500000e+01  
Nbhd        4.320000e+03  5.283000e+03  6.982000e+03  
Stories     1.000000e+00  2.000000e+00  4.000000e+01  
Year_Built  1.950000e+03  1.959000e+03  2.024000e+03  
Nr_of_rms   0.000000e+00  0.000000e+00  1.200000e+01  
Fin_sqft    1.344000e+03  1.943000e+03  3.787170e+05  
Units       1.000000e+00  1.000000e+00  7.810000e+02  
Bdrms       3.000000e+00  4.000000e+00  2.031000e+03  
Fbath       1.000000e+00  2.000000e+00  1.000000e+01  
Hbath       0.000000e+00  1.000000e+00  1.000000e+01  
PropertyID  2.226960e+05  8.567878e+05  8.814760e+05  
taxkey      3.220466e+09  4.901190e+09  7.160380e+09  
nbhd        3.060000e+03  4.660000e+03  4.270300e+04  
Rooms       6.000000e+00  8.000000e+00  7.000000e+01

### Daten visualisieren

In [694]:
''' 
# Erstellung eines Profilingreports
pr = ProfileReport(df, title = 'Property Sales') 
filename_pr = "../output/property_sales_pr.html" 
path_pr = os.path.abspath(filename_pr) 

pr.to_file(path_pr)  # ProfileReport als HTML speichern
webbrowser.open(f"file://{path_pr}")  # ProfileReport im Browser öffnen
'''

' \n# Erstellung eines Profilingreports\npr = ProfileReport(df, title = \'Property Sales\') \nfilename_pr = "../output/property_sales_pr.html" \npath_pr = os.path.abspath(filename_pr) \n\npr.to_file(path_pr)  # ProfileReport als HTML speichern\nwebbrowser.open(f"file://{path_pr}")  # ProfileReport im Browser öffnen\n'

In [695]:
# TODO Geo Daten anzeigen (wo wie verztilt sind)

In [696]:
''' 
from geopy.geocoders import Nominatim
import folium
from geopy.extra.rate_limiter import RateLimiter

df_geo = df[['Address']]
df_geo = df_geo.drop_duplicates()
df_geo = df_geo.dropna(subset=["Address"])
df_geo
'''


' \nfrom geopy.geocoders import Nominatim\nimport folium\nfrom geopy.extra.rate_limiter import RateLimiter\n\ndf_geo = df[[\'Address\']]\ndf_geo = df_geo.drop_duplicates()\ndf_geo = df_geo.dropna(subset=["Address"])\ndf_geo\n'

In [697]:
# Anzeige der kleinsten Einträge bei der Variable Year_Built, um den Schwellenwert für fehlerhafte Einträge zu identifizieren
print(df['Year_Built'].drop_duplicates().nsmallest(10).tolist())

# Anzeige der größten Einträge bei der Variable Bdrms, um den Schwellenwert für fehlerhafte Einträge zu identifizieren
print(df['Bdrms'].drop_duplicates().nlargest(10).tolist())


[0.0, 203.0, 206.0, 910.0, 1835.0, 1836.0, 1840.0, 1843.0, 1844.0, 1848.0]
[2031.0, 43.0, 32.0, 29.0, 28.0, 25.0, 21.0, 20.0, 18.0, 17.0]


## Datenaufbereitung

In [698]:
# Duplikate löschen
df = df.drop_duplicates()

In der vorherigen Dateneinsicht haben sich folgende Aufbereitsungsschritte für die einzelnen Variablen ergeben:

- PropertyID: Kann gelöscht werden, da diese Variable keine sinnvollen Informationen beinhaltet.
- PropType: Fehlende Einträge (NaN) werden auf "Unknown" gesetzt.
- taxkey: Wird gelöscht, da der Steuerschlüssel sich 
- Address: Fehlende Einträge (NaN) werden auf "Unknown" gesetzt.
- CondoProject: Fehlende Einträge (NaN) werden auf "None" gesetzt
- District
- nbhd
- Style: Fehlende Einträge (NaN) werden auf "Unknown" gesetzt.
- Extwall: Fehlende Einträge (NaN) werden auf "Unknown" gesetzt.
- Stories
- Year_Built: Einträge bei denen eine extreme niedrige Jahresanzahl (n < 1000) auftritt werden gelöscht.
- Rooms
- FinishedSqft
- Units
- Bdrms: Einträge bei denen eine extreme und unnatürliche Anzahl an Schlafzimmern (n > 100) auftrit werden gelöscht.
- Fbath
- Hbath
- Lotsize: Datentyp in Integer ändern.
- Sale_date
- Sale_price: Fehlende Einträge (NaN) werden gelöscht, da es sich bei dieser Variable um die Zielvariable handelt. Datentyp in Integer ändern.

CondoProject		84.185703
Nr_of_rms		55.793598
Nbhd		55.793598
Taxkey		55.793598
Fin_sqft		55.793598


In [699]:
df.loc[df['Year_Built'] < 1000, 'Year_Built'] = np.nan
df.loc[df['Bdrms'] > 100, 'Bdrms'] = np.nan


df['PropType'] = df['PropType'].fillna('Unknown')
df['Style'] = df['Style'].fillna('Unknown')
df['Extwall'] = df['Extwall'].fillna('Unknown')
df['Address'] = df['Address'].fillna('Unknown')
df['CondoProject'] = df['CondoProject'].fillna('None')

# NaN-Werte der Spalze Sale_price werden gelöscht, da es sich hierbei um die Zielvariable hadnelt
df = df.dropna(subset=['Sale_price'])

In [700]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 67531 entries, 0 to 67634
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PropType      67531 non-null  object 
 1   Taxkey        29859 non-null  float64
 2   Address       67531 non-null  object 
 3   CondoProject  67531 non-null  object 
 4   District      67529 non-null  float64
 5   Nbhd          29859 non-null  float64
 6   Style         67531 non-null  object 
 7   Extwall       67531 non-null  object 
 8   Stories       67345 non-null  float64
 9   Year_Built    66881 non-null  float64
 10  Nr_of_rms     29859 non-null  float64
 11  Fin_sqft      29859 non-null  float64
 12  Units         67531 non-null  float64
 13  Bdrms         65587 non-null  float64
 14  Fbath         66924 non-null  float64
 15  Hbath         62749 non-null  float64
 16  Lotsize       67530 non-null  object 
 17  Sale_date     67531 non-null  object 
 18  Sale_price    67531 non-null  o

In [701]:
df = df.drop(columns=['PropertyID', 'Taxkey','taxkey', 'nbhd', 'Rooms', 'FinishedSqft', 'Sale_date'])

In [702]:
 # Spalten die eig. ganzzahlig sind, werden auf das Datenformat Int geändert
for col in df.select_dtypes(include="float"):
    if (df[col].dropna() % 1 == 0).all():  # Prüfung, ob alle Werte auf .0 enden (ganzzahlig sind), dabei werden NaN Werte nicht betrachtet
        df[col] = df[col].astype("Int64")

# Datentyp von Lotsize in Int ändern
df['Lotsize'] = pd.to_numeric(df['Lotsize'].astype(str).str.replace(',', '', regex=False),errors='coerce').astype('Int64')
# Datentyp von Sale_price in Int ändern
df['Sale_price'] = pd.to_numeric(df['Sale_price'].astype(str).str.replace('[^\d]', '', regex=True), errors='coerce').astype('Int64')

# Datentyp von Sale_date auf Datetime ändern
#df["Sale_date"] = pd.to_datetime(df["Sale_date"], format="%Y-%m", errors='coerce').dt.to_period("M")

In [703]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 67531 entries, 0 to 67634
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PropType      67531 non-null  object 
 1   Address       67531 non-null  object 
 2   CondoProject  67531 non-null  object 
 3   District      67529 non-null  Int64  
 4   Nbhd          29859 non-null  Int64  
 5   Style         67531 non-null  object 
 6   Extwall       67531 non-null  object 
 7   Stories       67345 non-null  float64
 8   Year_Built    66881 non-null  Int64  
 9   Nr_of_rms     29859 non-null  Int64  
 10  Fin_sqft      29859 non-null  Int64  
 11  Units         67531 non-null  Int64  
 12  Bdrms         65587 non-null  Int64  
 13  Fbath         66924 non-null  Int64  
 14  Hbath         62749 non-null  Int64  
 15  Lotsize       67530 non-null  Int64  
 16  Sale_price    67531 non-null  Int64  
dtypes: Int64(11), float64(1), object(5)
memory usage: 10.0+ MB


In [704]:
list_cols_to_delete = ['PropertyID']
list_cols_location = ['Address', 'District', 'Nbhd']
target_var = 'Sale_price'

## Dateneinsicht Teil 2

In [705]:
df.describe().T

count            mean             std     min       25%  \
District    67529.0        7.958107        4.219026     1.0       4.0   
Nbhd        29859.0     3822.611005     1799.315293    40.0    2100.0   
Stories     67345.0        1.354963        0.606624     0.0       1.0   
Year_Built  66881.0     1943.799031       29.773495  1835.0    1923.0   
Nr_of_rms   29859.0        0.683546        1.641416     0.0       0.0   
Fin_sqft    29859.0     2535.138518     9264.489001     0.0    1063.0   
Units       67531.0        1.524307        8.114528     0.0       1.0   
Bdrms       65587.0         3.11022        1.579398     0.0       2.0   
Fbath       66924.0        1.375067        0.712695     0.0       1.0   
Hbath       62749.0        0.317057        0.513835     0.0       0.0   
Lotsize     67530.0     38877.51195  6785854.768201     0.0    3600.0   
Sale_price  67531.0  1347804.984185  6050470.026644     0.0  145000.0   

                 50%        75%           max  
District         8.0       11.0          15.0  
Nbhd          4320.0     5283.0        6982.0  
Stories          1.0        2.0          40.0  
Year_Built    1950.0     1959.0        2024.0  
Nr_of_rms        0.0        0.0          12.0  
Fin_sqft      1344.0     1942.0      378717.0  
Units            1.0        1.0         781.0  
Bdrms            3.0        4.0          43.0  
Fbath            1.0        2.0          10.0  
Hbath            0.0        1.0          10.0  
Lotsize       5040.0     6916.0  1703065320.0  
Sale_price  400000.0  1700000.0   950000000.0

In [706]:
categoricalVar = [col for col in df if df[col].dtype == 'object']

# Anzahl der einzigartigen Werte pro Spalte
df[categoricalVar].nunique().sort_values(ascending=False)

Address         51304
CondoProject      549
Style             400
Extwall            23
PropType            8
dtype: int64